Implementation of the XTypes type system

dennis-adlink opened the pull request:
This PR builds on PR #903 and contains the implementation of the XTypes type system in ddsi and ddsc. This replaces the current implementation of type identifiers and type discovery. The new implementation uses the Type Identifier and Type Objects as defined in the DDS XTypes specification for type discovery and assignability checking.

The commit message of the (rather large) commit fc75a9b contains a more detailed description of the changes, and also lists some of the main features that are not (fully) implemented yet.

dennis-adlink commented on src/core/ddsi/src/ddsi_typelib.c:
```diff
@@ -0,0 +1,507 @@
+struct ddsi_type * ddsi_type_ref_proxy (struct ddsi_domaingv *gv, const ddsi_typeinfo_t *type_info, ddsi_typeid_kind_t kind, const ddsi_guid_t *proxy_guid)
+{
+  assert (type_info);
+  assert (kind == DDSI_TYPEID_KIND_MINIMAL || kind == DDSI_TYPEID_KIND_COMPLETE);
```
The functions `ddsi_type_ref_local` and `ddsi_type_ref_proxy` are used to reference (and store if not an existing) the top-level type from a sertype (`..ref_local`) or from the type information in discovery data of a proxy endpoint (`.._ref_proxy`). A type identifier of a top-level type can't be fully descriptive, as the top-level type is a union or struct. The `kind` parameter is not taken from discovery data, but is a static value (in all cases both minimal and complete types are referenced, so I could combine both in a single function..)

dennis-adlink commented:
The update that I've just pushed to this PR branch contains fixes for the issues mentioned above (most of them squashed into fc75a9b), and in addition to that:
- bugfixes for issues found with fuzzer tests and manual testing
- refactoring of the `sertype->derive_sertype` interface (see commit 85cfe60)
- change the interface of the `dds_builtintopic_get_endpoint_typeid` api function and add the `dds_domain_get_typeobj` function (55ea999 and 4cc004a)
- fixed for the ddsi_type_id (and type_info/type_map) typedef/forward declaration issue (00ad424)
- replace `typeid` by `type_id` as the former is a reserved keyword in c++ (43cac30)

There are still some open issues, e.g. in mutable type assignability rules, an idlc type-object generator crash for specific idl and some other, which I will fix. So I'll leave this PR in draft state for now.

dennis-adlink commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -78,24 +78,21 @@ static dds_return_t dds_topic_status_validate (uint32_t mask)
-    // tid may be a null pointer but it may also be all-zero
-    ddsrt_free (tid);
+  ddsi_typeid_t *tid = ddsi_sertype_typeid (sertype, DDSI_TYPEID_KIND_COMPLETE);
+  if (ddsi_typeid_is_none (tid))
```
`dds_create_topic_impl` now checks that, in case there is type information in the sertype, it has both a minimal and complete type id set: https://github.com/dennis-adlink/cyclonedds/blob/95f76dfc5be1a6c8c8f1963da9f64f9be94a60a9/src/core/ddsc/src/dds_topic.c#L437

eboasson commented on src/core/ddsi/src/ddsi_plist.c:
```diff
@@ -678,6 +681,104 @@ static bool print_data_representation (char * __restrict *buf, size_t * __restri
+  ddsi_typeinfo_t const * const * x = deser_generic_src (src, &srcoff, alignof (ddsi_typeinfo_t *));
+
+  dds_ostream_t os = { .m_buffer = NULL, .m_index = 0, .m_size = 0, .m_xcdr_version = CDR_ENC_VERSION_2 };
+  if (bo == DDSRT_BOSEL_LE)
```
I'd find it more elegant if there'd be a `dds_stream_write_with_byte_order` or so. It only moves the cases here to another place, but it feels better to me.

dennis-adlink replied:
Fixed in adb9c5e

eboasson commented on src/core/ddsi/src/ddsi_plist.c:
```diff
@@ -678,6 +681,104 @@ static bool print_data_representation (char * __restrict *buf, size_t * __restri
+static dds_return_t valid_type_information (const void *src, size_t srcoff)
+{
+  ddsi_typeinfo_t const * const * x = deser_generic_src (src, &srcoff, alignof (ddsi_typeinfo_t *));
+  /* FIXME: add more checks for type ids */
```
Perhaps you can elaborate a bit, even if just in a GitHub comment?

dennis-adlink replied:
Sure, and because this validation is pretty straightforward I decided to implemented these checks:
https://github.com/dennis-adlink/cyclonedds/commit/eda0edc365775081ea8ebd312d7c7e410b1ae0f8

But when testing this, I noticed that the `valid` function is never called for this parameter (I think because it is defined as a `QP` and not a `PP`). I'm not sure what's the best way to implement this, can I add a call to `op.f.valid` in `init_one_parameter`? Or call the validate function from `deser_type_information`?

dennis-adlink replied:
After studying the code in `ddsi_plist.c` a bit more, I think this is the right approach for calling the `valid` function for parameters: 906b1041959aec899d935f0c977a0146828346e6 @eboasson could you please check if this is correct?

eboasson commented on src/core/ddsi/src/ddsi_typelookup.c:
```diff
@@ -236,43 +50,59 @@ static struct writer *get_typelookup_writer (const struct ddsi_domaingv *gv, uin
+  type->request_seqno++;
+  request.header.requestId.sequence_number.high = (int32_t) (type->request_seqno >> 32);
+  request.header.requestId.sequence_number.low = (uint32_t) type->request_seqno;
+  // FIXME: request.header.instanceName = ...
```
This one looks like it may need to be dealt with?

dennis-adlink replied:
This is added in ce4d331. The specs says it should be a 16-char string of the participant GUID, but the 16-byte GUID doesn't fit in so we'll include it as a 32-char string for now (an OMG xtypes spec issue is created for this: https://issues.omg.org/issues/DDSXTY14-53)

dennis-adlink commented on src/core/ddsi/src/ddsi_typelib.c:
```diff
@@ -0,0 +1,507 @@
+    for (uint32_t n = 0; dep_ids && n < dep_ids->_length; n++)
+    {
+      const ddsi_typeid_t *dep_type_id = &dep_ids->_buffer[n].type_id;
+      if (ddsi_typeid_compare (&type->xt.id, dep_type_id) && !type_has_dep (type, dep_type_id))
```
You're right, and although the number of dependent types will typically not be very large and this is only used when a new endpoint is created or discovered, a better approach to keep track of type dependencies is needed. I'll add a fixme comment here.

dennis-adlink commented on src/core/ddsi/src/ddsi_typewrap.c:
```diff
@@ -0,0 +1,1831 @@
+      }
+      break;
+    default:
+      abort (); /* not supported */
```
The type object that is passed to this function can either be from a type-map in a topic descriptor for a locally created topic, but also from a remote node (type-lookup reply). Especially the latter needs proper error handling, so I've replaced this assert by returning DDS_RETCODE_UNSUPPORTED, and error handling at places where `ddsi_xt_type_add_typeobj` is used.

dennis-adlink commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -78,24 +78,18 @@ static dds_return_t dds_topic_status_validate (uint32_t mask)
-    ddsrt_free (tid);
+  ddsi_typeid_t *type_id = ddsi_sertype_typeid (sertype, DDSI_TYPEID_KIND_COMPLETE);
+  if (ddsi_typeid_is_none (type_id))
     return NULL;
```
The type id is returned by a sertype implementation, so it's dangerous to assume that the `DDS_XTypes_TK_NONE` case cannot occur. I'll replaced the check by `ddsi_typeid_is_none (type_id) || !ddsi_typeid_is_hash (type_id)` (because a sertype's typeid must be a hashed id) and free the type id also in case it is null or non-hash.

dennis-adlink commented on src/core/ddsc/src/dds_topic.c:
```diff
@@ -345,48 +340,47 @@ static dds_entity_t create_topic_pp_locked (struct dds_participant *pp, struct d
 
-      m = ddsrt_malloc (sizeof (*m));
-      m->type_id = tid;
+      m = dds_alloc (sizeof (*m));
```
My idea was to allocate objects of types that are defined in DDSC with `dds_alloc` (in this case: `struct ktopic_type_guid`, `dds_ktopic`), including their members (unless the member type is defined in ddsi, e.g. `ddsi_typeid_t` which is allocated in `sertype_default`). For types defined in DDSI, I've used `ddsrt_alloc`/`ddsrt_free`. For now I'll make this approach a bit more consistent in this file by replacing a few alloc/free instances in code that changed in this PR. 

But I'm not sure if this is the right approach, I also don't mind to change all to `dds_alloc`/`dds_free` if you prefer that.

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/q_entity.h:
```diff
@@ -197,6 +197,13 @@ struct nn_rsample_info;
 struct ddsi_tkmap_instance;
 
+typedef struct ddsi_type_pair {
+#ifdef DDS_HAS_TYPE_DISCOVERY
```
Fixed in ff3d9d7

dennis-adlink commented on src/core/ddsc/src/dds_domain.c:
```diff
@@ -514,4 +519,76 @@ dds_return_t dds_domain_resolve_type (dds_entity_t entity, unsigned char *type_i
   return rc;
 }
+
+dds_return_t dds_domain_get_typeobj(
```
You're right, I've refactored these functions in 046d067

eboasson commented:
@dennis-adlink it looks good to me. I'm sure that somewhere deep in depths of this there'll be some overlooked issue, but I don't believe we're going to improve much upon it while keeping it separate and so I think it is time to merge it. (And it would unblock a number of other things besides, like in the python and C++ bindings.)

If you would be kind enough to rebase it (to me it looks trivial) and force-push it, I'll have one final look at the changes introduced by the rebasing itself. Perhaps it is stupid to hope to merge such a big piece of work in *just before* the holidays, but I would consider it a nice little present 🙂

dennis-adlink commented:
@eboasson I've rebased this branch on the changes in #1082, which are required to make the python fuzzer tests pass. And I've fixed a few minor issues after the rebase, in commit df4af91 and 6c1850e. 

@thijsmie: the function name change of `dds_get_typeobj` in commit 70b7d34 needs a rename in the XTypes enabled version of pysertype.c in the Python binding

thijsmie commented:
🥳
